lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7954,14 +7988,32 @@ CustomRefCountingOperationResult CustomRefCountingOperation::evaluate(
                                  : "release:";
 
   auto decl = cast<clang::RecordDecl>(swiftDecl->getClangDecl());
-  if (!decl->hasAttrs())
+
+  auto [declHasAReferenceTypeBase, baseRefDecl] =
+      getFRTInfoFromBaseDecls(ctx, evaluator, decl);
+
+  if (!decl->hasAttrs() && declHasAReferenceTypeBase &&
+      !baseRefDecl->hasAttrs())
     return {CustomRefCountingOperationResult::noAttribute, nullptr, ""};
 
   llvm::SmallVector<const clang::SwiftAttrAttr *, 1> retainReleaseAttrs;
-  for (auto *attr : decl->getAttrs()) {
-    if (auto swiftAttr = llvm::dyn_cast<clang::SwiftAttrAttr>(attr)) {
-      if (swiftAttr->getAttribute().starts_with(operationStr)) {
-        retainReleaseAttrs.push_back(swiftAttr);
+  if (decl->hasAttrs()) {
+    for (auto *attr : decl->getAttrs()) {
```

Comment:
The code is duplicated here and below. We could create a lambda and call it with `decl` and `baseRefDecl` to deduplicate.